core/sr-io/src/lib.rs

```diff
@@ -279,7 +279,7 @@ export_api! {
 		/// Submit transaction to the pool.
 		///
 		/// The transaction will end up in the pool.
-		fn submit_transaction<T: codec::Encode>(data: &T) -> Result<(), ()>;
+		fn submit_transaction(data: Vec<u8>) -> Result<(), ()>;
```

Comment:
This is a breaking change, does it also mean we won't be able to do generic functions in the future on the `sr-io` level?
I was actually thinking of making it properly typed (i.e. we expect `Block::Extrinsic` here).

Reply:
That is no breaking change (in the sense that it does not breaks any chain). We don't change the `ext_submit_transaction` in this case.

The problem is that you can not generate c functions that take generics. In the next pr, I will start to clean up the usage of `generics` in the `sr-io` interface functions. We really should move this into the structs. 

We have for example `ordered_trie_root` and expose the function `ext_blake2_256_enumerated_trie_root`. The original name should directly be `blake2_256_ordered_trie_root` and then we have a trait in `sr-primitives` that provides the ordered trie root function and the implementation for `Blake2Hasher` calls the correct function.

`submit_transaction` should follow the same pattern and the trait should be implemented directly for the extrinsic types.